### PR TITLE
Use weak password hashing for testing

### LINKED ISSUE #164: Use weak password hashing for testing in course-discovery
We can improve the total time of the test suite by choosing a weak password hasher for the test environment only. 

Ref: https://docs.djangoproject.com/en/4.1/topics/testing/overview/#password-hashing

### PR DESCRIPTION
**Pull request by joelantoniov**
## Description

This pull request addresses the issue regarding slow test suite performance caused by strong password hashing during testing. To improve the test suite runtime, I have implemented a solution that uses weak password hashing specifically for testing purposes, as recommended by the Django documentation: https://docs.djangoproject.com/en/4.1/topics/testing/overview/#password-hashing

## Changes Made

- Created a new test file `course_discovery/settings/test_weak_password.py` to implement weak password hashing for testing.
- Modified the test settings to use the weak password hashing configuration only during testing.
- Added a test case to verify that weak password hashing is applied during testing.

This change has been tested locally and successfully improves the test suite runtime while ensuring the security of strong password hashing in production.

Fixes: https://github.com/openedx/public-engineering/issues/164

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @joelantoniov!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by joelantoniov** [comment]
Hi, @e0d! In order to avoid any issues with the previously pushed and amended commits, I've created this new pull request to update my local repository. I already  amended the commits and  I created this new pull request. My apologies for any confusion!

**Comment by e0d** [comment]
Your branch is behind the base.  I've pulled in changes from master as a merge commit which will update your branch and cause the tests to be re-run.

**Comment by joelantoniov** [comment]
> **Lint Commit Messages / commitlint / commitlint (pull_request) ** Failing after 16s

@e0d I really do not understand the Lint Commit Messages problems, because I am using the right style...

**Comment by e0d** [comment]
The issue seems to be with this commit, https://github.com/openedx/course-discovery/pull/4070/commits/e70fb0e81725296f19b5ae3d698612af2172623e, there are additional meta-data requirements for *feat I believe.

**Comment by joelantoniov** [comment]
> The issue seems to be with this commit, [e70fb0e](https://github.com/openedx/course-discovery/commit/e70fb0e81725296f19b5ae3d698612af2172623e), there are additional meta-data requirements for *feat I believe.

@e0d , I already changed the commit message.

**Comment by mphilbrick211** [comment]
Hi @joelantoniov! Just flagging that there's a failing test.

**Comment by joelantoniov** [comment]
> Hi @joelantoniov! Just flagging that there's a failing test.

Hi, @mphilbrick211 . Yes, I do not understand why, because this very same was not failing in the previous running. Besides, I only changed the message commit and I did not touch the code, weird. Anywyas, since this test has not the required flag, would means it could proceed with the merge?

**Comment by mphilbrick211** [comment]
Hi @AliAdnanSohail! Are you able to help @joelantoniov with his question?

**Comment by AliAdnanSohail** [comment]
Hi @joelantoniov - Please rebase your branch with latest master. It seems the failing test doesn't belongs to your change. I'll look into it in detail if test will not pass after rebase

Thanks

File: course_discovery/settings/base.py [hunk 1]
```diff
@@ -316,6 +316,19 @@
 # Request the user's permissions in the ID token
 EXTRA_SCOPE = ['permissions']
 
+PASSWORD_HASHERS = [
+    'django.contrib.auth.hashers.MD5PasswordHasher',  # Use MD5 hasher for testing
```
Comment (DawoudSheraz) [inline course_discovery/settings/base.py]: If it is to be used for testing, shouldn't it be in test.py?

File: course_discovery/settings/base.py [hunk 2]
```diff
@@ -316,6 +316,19 @@
...
+]
+
+AUTH_PASSWORD_VALIDATORS = [
+    {
+        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
+    },
+    {
+        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
+    },
+]
```
Comment (DawoudSheraz) [inline course_discovery/settings/base.py]: Why are these needed?

File: course_discovery/settings/test_weak_password.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#! /usr/bin/env python
```
Comment (DawoudSheraz) [inline course_discovery/settings/test_weak_password.py]: Why is this file in settings directory?

File: course_discovery/settings/test_weak_password.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+# -*- coding: utf-8 -*-
+
+from django.contrib.auth.hashers import make_password
+from django.test import TestCase
+
+#from core.models import User
```
Comment (DawoudSheraz) [inline course_discovery/settings/test_weak_password.py]: nit: commented code.

File: course_discovery/settings/base.py (hunk 1 above)
Comment (joelantoniov) [reply course_discovery/settings/base.py]: To improve the test suite runtime, I have implemented a solution that uses weak password hashing specifically for testing purposes, as recommended by the Django documentation: https://docs.djangoproject.com/en/4.1/topics/testing/overview/#password-hashing

So, I created a new test file course_discovery/settings/test_weak_password.py to implement weak password hashing for testing. I can move it to test.py if it is not right.

File: course_discovery/settings/base.py (hunk 2 above)
Comment (joelantoniov) [reply course_discovery/settings/base.py]: These password validators are just one aspect of that security posture, and they help protect user accounts from easily guessable or weak passwords. I will move this also to test.py

File: course_discovery/settings/base.py (hunk 1 above)
Comment (AliAdnanSohail) [reply course_discovery/settings/base.py]: By adding in base.py these settings will be applied on all environment including development, production and testing. If you need something just in testing environment then it must be added in test.py. Otherwise it will impact on other environments as well. So please move this to test.py only

**Comment by openedx-webhooks** [comment]
@joelantoniov Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/course-discovery&entry.867055334=https://github.com/openedx/course-discovery/pull/4070&entry.1484655318=https://github.com/joelantoniov&entry.752974735=2023-08-29+16:31&entry.1917517419=2023-09-26+15:27&entry.2133058324=No) so we can improve your experience in the future.

**Comment by openedx-webhooks** [comment]
@joelantoniov Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/course-discovery&entry.867055334=https://github.com/openedx/course-discovery/pull/4070&entry.1484655318=https://github.com/joelantoniov&entry.752974735=2023-08-29+16:31&entry.1917517419=2023-09-26+15:27&entry.2133058324=No) so we can improve your experience in the future.
